### PR TITLE
[Core] Support launching a new cluster from an existing cluster's disk

### LINKED ISSUE #2050: Support migration of a stopped cluster to another region
When a cluster is STOPPED and the resource of that cluster is not available in that region/zone, the user will not be able to restart the cluster and it would be hard to migrate to another region manually as the cluster may be a dev machine.

To mitigate the problem, we should provide a way to migrate the stopped cluster across regions or even clouds.

The following is a script that can be used to migrate a cluster from one region to another on AWS:

```
# Variables need to be changed:
REGION='us-east-1'
TARGET_REGION='us-west-2'
CLUSTER_NAME='test-image-transfer'
AMI_NAME='image-transfer-ami0'

echo "Finding instance ID for cluster $CLUSTER_NAME in region $REGION..."
INSTANCE_ID=`aws ec2 describe-instances --region $REGION --filters Name=tag:ray-cluster-name,Values=$CLUSTER_NAME --query Reservations[].Instances[].InstanceId --output text`

echo "Creating AMI from instance $INSTANCE_ID in region $REGION..."
IMAGE_ID=`aws ec2 create-image --instance-id $INSTANCE_ID --name $AMI_NAME --region $REGION --output text`

echo Waiting for AMI to be available... This may take a few minutes.
aws ec2 wait image-available --image-ids $IMAGE_ID --region $REGION

echo "Copying AMI $IMAGE_ID from region $REGION to region $TARGET_REGION..."
NEW_IMAGE_ID=`aws ec2 copy-image --name $AMI_NAME --source-image-id $IMAGE_ID --source-region $REGION --region $TARGET_REGION --output text`

echo Waiting for AMI to be available... This may take a few minutes.
aws ec2 wait image-available --image-ids $NEW_IMAGE_ID --region $TARGET_REGION

echo "New AMI ID in $TARGET_REGION: $NEW_IMAGE_ID"

echo "You can now terminate the old cluster and launch a new cluster with the new AMI:"
echo "  sky down $CLUSTER_NAME -y"
echo "  sky launch -c $CLUSTER_NAME --cloud aws --region $TARGET_REGION -y"

```

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #2050. This PR only supports AWS and GCP as an experimental feature and unblocks the users.

~~It is blocked by #2087~~

`sky launch --clone-disk-from existing-cluster -c new-cluster --cloud aws --region us-west-2 task.yaml`

A problem with the image copying is that the copying can take significant of time.

TODOs:
- [x] Support GCP
- [x] Clean up the image after the target cluster being tear down.
- [x] docs for additional permissions required

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - AWS (with both admin and minimal permission with the new permissions):
  - [x] `sky launch -c min --cloud aws --cpus 2 --region us-east-1`; `ssh min`; `echo 'hi' > user_file.txt`
  - [x] `sky launch --clone min -c new-min --cloud aws --region us-west-2`
  - [x] `sky launch --clone min -c new-min2`
  - [x] `sky down new-min new-min2`
  - GCP:
  - [x] `sky launch -c min --cloud gcp --cpus 2 --region us-central1`; `ssh min`; `echo 'hi' > user_file.txt`
  - [x] `sky launch --clone min -c new-min --cloud gcp --region us-east1 --disk-size 50`
  - [x] `sky launch --clone min -c new-min --cloud gcp --region us-east1`
  - [x] `sky launch --clone min -c new-min2`
  - [x] `sky down new-min new-min2`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/cli.py [hunk 1]
```diff
@@ -1259,6 +1271,15 @@ def cli():
...
               help='Skip setup phase when (re-)launching cluster.')
+@click.option(
+    '--clone-disk-from',
+    '--clone',
+    default=None,
+    type=str,
+    **_get_shell_complete_args(_complete_cluster_name),
+    help=('[Experimental] Clone disk from an existing cluster to launch '
+          'a new one. This is useful when the new cluster needs to have '
+          'the same data on the boot disk as the existing cluster.'))
```
Comment (concretevitamin) [inline sky/cli.py:1282]:
```suggestion
          'the same data on the boot disk as an existing cluster.'))
```

File: sky/backends/backend_utils.py [hunk 2]
```diff
@@ -1703,6 +1710,81 @@ def _query_cluster_status_via_cloud_api(
...
 
+def check_clone_disk_and_override_task(
+    cluster_name: str, target_cluster_name: Optional[str], task: 'task_lib.Task'
+) -> Tuple['task_lib.Task', 'cloud_vm_ray_backend.CloudVmRayResourceHandle']:
+    """Check if the task is compatible to clone disk from the source cluster.
+
+    Args:
+        cluster_name: The name of the cluster to clone disk from.
+        task: The task to check.
+        backend: The backend to use.
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1721]: Order / args seem wrong.

File: sky/backends/backend_utils.py [hunk 3]
```diff
@@ -1703,6 +1710,81 @@ def _query_cluster_status_via_cloud_api(
...
+    source_cluster_status, handle = refresh_cluster_status_handle(cluster_name)
+
+    if target_cluster_name is not None:
+        target_cluster_status, _ = refresh_cluster_status_handle(
+            target_cluster_name)
+        if target_cluster_status is not None:
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.NotSupportedError(
+                    'The target cluster already exists. Cloning disk is only '
+                    'supported for creating a new cluster.')
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1739]:
```suggestion
                    'supported when creating a new cluster.')
```

File: sky/backends/backend_utils.py [hunk 4]
```diff
@@ -1703,6 +1710,81 @@ def _query_cluster_status_via_cloud_api(
...
+
+    if source_cluster_status is None:
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1741]: Move to L1731?

File: sky/backends/backend_utils.py [hunk 5]
```diff
@@ -1703,6 +1710,81 @@ def _query_cluster_status_via_cloud_api(
...
+    if not isinstance(handle, backends.CloudVmRayResourceHandle):
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.NotSupportedError(
+                f'Cannot clone disk from a non-cloud cluster {cluster_name!r}.')
+
+    if source_cluster_status != status_lib.ClusterStatus.STOPPED:
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.NotSupportedError(
+                'Cannot clone disk from a RUNNING cluster. Please stop the '
+                f'cluster first: sky stop {cluster_name}.')
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1755]: Likewise, move to earlier?

File: sky/backends/backend_utils.py [hunk 6]
```diff
@@ -1703,6 +1710,81 @@ def _query_cluster_status_via_cloud_api(
...
+
+    if not isinstance(handle, backends.CloudVmRayResourceHandle):
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.NotSupportedError(
+                f'Cannot clone disk from a non-cloud cluster {cluster_name!r}.')
+
+    if source_cluster_status != status_lib.ClusterStatus.STOPPED:
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.NotSupportedError(
+                'Cannot clone disk from a RUNNING cluster. Please stop the '
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1754]:
```suggestion
                f'Cannot clone disk from a {source_cluster_status} cluster. Please stop the '
```

File: sky/backends/backend_utils.py [hunk 7]
```diff
@@ -1703,6 +1710,81 @@ def _query_cluster_status_via_cloud_api(
...
+
+    if override_param:
+        logger.info(
+            f'No cloud/region specified for the task. Using the same region '
+            f'as source cluster {cluster_name!r}: '
+            f'{handle.launched_resources.cloud}'
+            f'({handle.launched_resources.region}).')
+        task_resources = task_resources.copy(**override_param)
+        task.set_resources({task_resources})
+        task.best_resources = None
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1784]: Why is this necessary?

File: sky/cli.py [hunk 8]
```diff
@@ -725,12 +725,21 @@ def _launch_with_confirm(
...
     """Launch a cluster with a Task."""
-    with sky.Dag() as dag:
-        dag.add(task)
     if cluster is None:
         cluster = backend_utils.generate_cluster_name()
+
+    clone_source_str = ''
+    if clone_disk_from is not None:
+        clone_source_str = f' from the disk of {clone_disk_from!r}'
+        task, _ = backend_utils.check_clone_disk_and_override_task(
```
Comment (concretevitamin) [inline sky/cli.py:737]: From the callsite not sure what this func does. How about "check_can_clone_disk_and_override_task"?

File: docs/source/cloud-setup/cloud-permissions.rst
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions.rst:1]: remnant?

File: sky/clouds/cloud.py [hunk 9]
```diff
@@ -27,6 +27,7 @@ class CloudImplementationFeatures(enum.Enum):
     STOP = 'stop'
     AUTOSTOP = 'autostop'
     MULTI_NODE = 'multi-node'
+    CLONE_DISK_FROM_CLUSTER = 'migrate-disk'
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:30]: Best to keep key & val in sync. Change one of them?

File: sky/execution.py [hunk 10]
```diff
@@ -424,6 +429,56 @@ def launch(
     entrypoint = task
     backend_utils.check_cluster_name_not_reserved(cluster_name,
                                                   operation_str='sky.launch')
+
+    if clone_disk_from:
```
Comment (concretevitamin) [inline sky/execution.py:433]: is not None

Comment (concretevitamin) [inline sky/execution.py:433]: Wondering if this is best to move into _execute(), perhaps as a Stage. Pros and cons? It seems odd to add a special case treatment here.

File: sky/clouds/cloud.py [hunk 11]
```diff
@@ -485,5 +486,25 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
...
+                         target_zone: Optional[str]) -> str:
+        """Move an image if the target region cannot find it.
+
+        Returns: the image ID in the target region.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def delete_image(cls, image_id: str, region: Optional[str]) -> None:
+        raise NotImplementedError
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:507]: We should add docstrs to all 3 methods, and:

- Say it's used by `sky launch --clone-disk-from`
- Say if it's ok for a cloud impl to leave these out. If so, they should add CloudImplementationFeatures.xx to the non supported list.

File: sky/clouds/cloud.py [hunk 12]
```diff
@@ -485,5 +486,25 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
         """
         raise NotImplementedError
 
+    @classmethod
+    def create_image_from_cluster(cls, name: str, tag_filters: Dict[str, str],
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:490]: API, just from reading the iface
- name -> image_name (or cluster_name)?
- tag_filters: Dict[str, str] -> seems like backend knowledge leaked to here?

File: sky/clouds/cloud.py [hunk 13]
```diff
@@ -485,5 +486,25 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
...
+                                  region: Optional[str],
+                                  zone: Optional[str]) -> str:
+        raise NotImplementedError
+
+    @classmethod
+    def maybe_move_image(cls, image_name: str, source_region: str,
+                         target_region: str, source_zone: Optional[str],
+                         target_zone: Optional[str]) -> str:
+        """Move an image if the target region cannot find it.
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:499]: Not sure what this means/if it captures the true intent?

My understanding is GCP's images are global so the "maybe move" part makes sense, but the "if the target region cannot find it" part doesn't?

File: sky/utils/subprocess_utils.py [hunk 14]
```diff
@@ -117,3 +120,37 @@ def kill_children_processes(first_pid_to_kill: Optional[int] = None,
         except psutil.NoSuchProcess:
             # The child process may have already been terminated.
             pass
+
+
+def run_and_retry(
```
Comment (concretevitamin) [inline sky/utils/subprocess_utils.py:125]: nit: docstr

**Review by concretevitamin — COMMENTED** [review]
Awesome @Michaelvll! Leaving some comments first, only a partial pass.

Some  UX notes:

1. I tried to pass a smaller disk size than an existing cluster
```
» sky launch --clone dbg -c smoke --region us-west2 --cloud gcp --disk-size 50 --cpus 4
...
Launching a new cluster 'smoke' from the disk of 'dbg'. Proceed? [Y/n]:
<spinner -- creating image>
Image 'projects/skypilot-xxxx/global/images/skypilot-dbg-xxxxxxxx' for 'dbg' created successfully on GCP. Overriding task's image_id.
ValueError: Image 'projects/skypilot-xxxx/global/images/skypilot-dbg-xxxxxxxx' is 256.0GB, which is larger than the specified disk_size: 50 GB. Please specify a larger disk_size to use this image.
```
The error shows up only after the image creation (which is now leaked). Can we pre-check?

Btw, this also makes me wonder what `_is_image_managed` means. The above ended up with an error - did we remove the auto-created image managed by us?

2. A successful clone:
```

Launching a new cluster 'smoke' from the disk of 'dbg'. Proceed? [Y/n]:
Image 'projects/skypilot-375900/global/images/skypilot-dbg-1687454357' for 'dbg' created successfully on GCP. Overriding task's image_id.
```
The second line probably should have a logging prefix, otherwise easy to miss. Also, green the line?

File: sky/execution.py (hunk 10 above)
Comment (Michaelvll) [reply sky/execution.py]: I don't have a strong preference for this, as this is only used by `launch`, but I moved it to the `_execution`, maybe to make it easier to read.

File: sky/utils/subprocess_utils.py (hunk 14 above)
Comment (Michaelvll) [reply sky/utils/subprocess_utils.py:125]: Done. Thanks!

**Comment by concretevitamin** [comment]
An odd behavior - I think I did the following:
```
sky launch --clone dbg -c smoke --region us-west2 --cloud gcp --cpus 4  # Success.
sky down dbg
```
A day after, 
```
sky cpunode -c smoke -i120
...
ValueError: Image 'projects/skypilot-12345/global/images/skypilot-dbg-xxxxx54357' not found in GCP.
```
On GCP console the image is indeed not there. Could it be `sky down dbg` deleted the image?

**Comment by Michaelvll** [comment]
Good catch! It was because we called the `post_teardown_cleanup` during the status refresh, which previously will delete image when the cluster is stopped. We should not delete the image when stopping the cluster. Fixed in the latest commit.

File: sky/execution.py [hunk 15]
```diff
@@ -228,31 +280,34 @@ def _execute(
...
+        bold = colorama.Style.BRIGHT
+        reset = colorama.Style.RESET_ALL
+        logger.info(
+            f'{yellow}Launching an unmanaged spot task, which does not '
+            f'automatically recover from preemptions.{reset}\n{yellow}To '
+            'get automatic recovery, use managed spot instead: '
+            f'{reset}{bold}sky spot launch{reset} {yellow}or{reset} '
+            f'{bold}sky.spot_launch(){reset}.')
+
+    if Stage.CLONE_DISK in stages:
```
Comment (concretevitamin) [inline sky/execution.py:295]: Seems like the 3 if clauses should be placed under an "if not cluster_exists:". Not feeling strongly, just felt that may be more readable.

File: sky/resources.py [hunk 16]
```diff
@@ -59,12 +45,78 @@ def __init__(
...
+        disk_size: Optional[int] = None,
         disk_tier: Optional[Literal['high', 'medium', 'low']] = None,
+        # Internal use only.
+        _is_image_managed: Optional[bool] = None,
     ):
+        """Initialize a Resources object.
+
+        All fields are optional.  ``Resources.is_launchable`` decides whether
+        the Resources is fully specified to launch an instance.
+
```
Comment (concretevitamin) [inline sky/resources.py:60]:
```suggestion
```

File: sky/resources.py [hunk 17]
```diff
@@ -285,6 +338,10 @@ def image_id(self) -> Optional[Dict[str, str]]:
     def disk_tier(self) -> str:
         return self._disk_tier
 
+    @property
+    def is_image_managed(self) -> Optional[str]:
```
Comment (concretevitamin) [inline sky/resources.py:342]:
```suggestion
    def is_image_managed(self) -> Optional[bool]:
```

File: sky/backends/cloud_vm_ray_backend.py [hunk 18]
```diff
@@ -3547,6 +3547,23 @@ def post_teardown_cleanup(self,
                             cluster_name=handle.cluster_name,
                             stdout=tpu_stdout,
                             stderr=tpu_stderr))
+        if handle.launched_resources.is_image_managed and terminate:
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3550]:
```suggestion
        if handle.launched_resources.is_image_managed is True and terminate:
```

Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3550]: Q: say I clone `old` into `new`. Then I run `sky stop new` and `sky launch -c new <some other resource args>`. Does it override its `handle.launched_resources.is_image_managed`?

File: sky/backends/cloud_vm_ray_backend.py [hunk 19]
```diff
@@ -3547,6 +3547,23 @@ def post_teardown_cleanup(self,
...
+            # Delete the image when terminating the cluster and the image is
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3551]:
```suggestion
            # Delete the image when terminating a "cloned" cluster, i.e., whose image is
```

File: sky/backends/cloud_vm_ray_backend.py [hunk 20]
```diff
@@ -3547,6 +3547,23 @@ def post_teardown_cleanup(self,
...
+            cluster_cloud = cluster_resources.cloud
+            image_dict = cluster_resources.image_id
+            assert cluster_cloud is not None, cluster_resources
+            assert image_dict is not None and len(image_dict) == 1
+            image_id = list(image_dict.values())[0]
+            try:
+                cluster_cloud.delete_image(image_id,
+                                           handle.launched_resources.region)
+            except exceptions.CommandError as e:
+                logger.debug(
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3564]: warn? "Failed to delete cloned image ... Please remove it manually to avoid image leakage. Details: ..."

File: sky/clouds/cloud.py [hunk 21]
```diff
@@ -485,5 +486,45 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
         """
         raise NotImplementedError
 
+    # === Image related methods ===
+    # These three tmethods are used to create, move and delete images. They
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:490]: tmethods -> typo

File: sky/clouds/gcp.py [hunk 22]
```diff
@@ -813,3 +817,91 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
...
+            instance_name_cmd,
+            error_msg=f'Failed to get instance name for {cluster_name!r}',
+            stderr=stderr,
+            stream_logs=True)
+        instance_names = json.loads(stdout)
+        if len(instance_names) != 1:
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.NotSupportedError(
+                    'Only support creating image from single '
+                    f' instance, but got multiple instances: {instance_names}')
```
Comment (concretevitamin) [inline sky/clouds/gcp.py:847]:
```suggestion
                    f' instance, but got: {instance_names}')
```
(can be 0?)

File: sky/clouds/aws.py [hunk 23]
```diff
@@ -695,6 +695,26 @@ def _get_disk_specs(cls, disk_tier: Optional[str]) -> Dict[str, Any]:
             'custom_disk_perf': tier != 'low',
         }
 
+    @classmethod
+    def _query_instance_property_and_retry(
```
Comment (concretevitamin) [inline sky/clouds/aws.py:699]: _with_retries?

File: sky/clouds/aws.py [hunk 24]
```diff
@@ -711,33 +731,9 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
...
-                    'Unable to locate credentials. You can configure credentials by '
-                    'running "aws configure"' in stdout + stderr):
-                retry_cnt += 1
-                time.sleep(random.uniform(0, 1) * 2)
-                continue
-            else:
-                break
+        assert region is not None, (tag_filters, region)
+        returncode, stdout, stderr = cls._query_instance_property_and_retry(
+            tag_filters, region, query='Reservations[].Instances[].State.Name')
```
Comment (concretevitamin) [inline sky/clouds/aws.py:736]: Seems like previous version has `"` around this query string. Is it needed?

File: sky/clouds/aws.py [hunk 25]
```diff
@@ -753,3 +749,134 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
...
+
+        subprocess_utils.handle_returncode(
+            returncode,
+            '',
+            error_msg='Failed to find the source cluster on AWS.',
+            stderr=stderr,
+            stream_logs=False)
+
+        instance_ids = json.loads(stdout.strip())
+        if len(instance_ids) != 1:
```
Comment (concretevitamin) [inline sky/clouds/aws.py:772]: Can be 0?

File: sky/utils/subprocess_utils.py [hunk 26]
```diff
@@ -117,3 +120,49 @@ def kill_children_processes(first_pid_to_kill: Optional[int] = None,
...
```
Comment (concretevitamin) [inline sky/utils/subprocess_utils.py:125]: run_with_retries?

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll! Did a pass, some comments.

File: sky/backends/backend_utils.py [hunk 27]
```diff
@@ -1752,15 +1754,19 @@ def check_can_clone_disk_and_override_task(
...
     task_resources = list(task.resources)[0]
     if handle.launched_resources.disk_size > task_resources.disk_size:
         # The target cluster's disk should be at least as large as the source.
         with ux_utils.print_exception_no_traceback():
+            target_cluster_name_str = f' {target_cluster_name!r}'
+            if target_cluster_name is None:
+                target_cluster_name_str = ''
             raise exceptions.NotSupportedError(
-                f'The target cluster {target_cluster_name!r} should have a disk size '
+                f'The target cluster {target_cluster_name_str} should have a disk size '
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1769]:
```suggestion
                f'The target cluster{target_cluster_name_str} should have a disk size '
```

**Review by concretevitamin — APPROVED** [review]
Thanks for solving a longstanding issue from GPU unavailability @Michaelvll!

File: docs/source/cloud-setup/cloud-permissions/aws.rst [hunk 28]
```diff
@@ -103,8 +103,22 @@ AWS accounts can be attached with a policy that limits the permissions of the ac
         ]
     }
 
-4. Click **Next: Tags** and follow the instructions to finish creating the policy. You can give the policy a descriptive name, such as ``minimal-skypilot-policy``.
-5. Go back to the previous window and click on the refresh button, and you can now search for the policy you just created.
+4. **Optional**: To allow ``--clone-disk-from`` to work, you need to add the following permissions to the policy above as well.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply docs/source/cloud-setup/cloud-permissions/aws.rst]: Good point! I tends to have it in the doc, as this doc page is not target for normal users but for admins, i.e., there is limited downside/burden this new entry will bring to a normal user. And it is easier to point our pilot user to this page, when they want to use this feature.

**Review by Michaelvll — COMMENTED** [review]
Tested:
- [x] `pytest tests/test_smoke.py::test_clone_disk_aws --aws`
- [x] `pytest tests/test_smoke.py::test_clone_disk_gcp`
